core/src/test/java/io/ably/lib/test/realtime/RealtimeAuthTest.java

```diff
@@ -956,4 +963,92 @@ public Object getTokenRequest(Auth.TokenParams params) throws AblyException {
         }
     }
 
+    /**
+     * Verify that instance of ClientOptions cannot change AblyRealtime options once it is provided to constructor
+     */
+    @Test
+    public void auth_client_options_immutable() {
+        try {
+            /* create token with clientId */
+            ClientOptions optsForToken = createOptions(testVars.keys[0].keyStr);
+            optsForToken.clientId = "token_clientId";
+            AblyBase<PushBase, Platform, RestChannelBase> ablyForToken = createAblyRest(optsForToken);
+            TokenDetails tokenDetails = ablyForToken.auth.requestToken(null, null);
+
+            /* create ably realtime */
+            ClientOptions opts = new ClientOptions();
+            opts.clientId = null;
+            opts.token = tokenDetails.token;
+            opts.autoConnect = false;
+            opts.headers = new HashMap<>();
+            opts.headers.put("old_key", "old_value");
+            opts.tokenDetails = new TokenDetails("my_old_details_token");
+            opts.logHandler = new Log.DefaultHandler();
+            opts.authCallback = new Auth.TokenCallback() {
+                @Override
+                public Object getTokenRequest(Auth.TokenParams params) {
+                    return null;
+                }
+            };
+            opts.authHeaders = new Param[1];
+            opts.authHeaders[0] = new Param("old_key", "old_key");
+            opts.proxy = new ProxyOptions();
+            opts.proxy.host = "https://ably.com";
+            opts.proxy.port = 8080;
+            AblyRealtimeBase<PushBase, Platform, RealtimeChannelBase> ablyRealtime = createAblyRealtime(opts);
+
+            assertEquals("clientId should be equal before change", opts.clientId, ablyRealtime.options.clientId);
+            opts.clientId = "my_new_clientId";
+            assertNotEquals("clientId should not be equal after change", opts.clientId, ablyRealtime.options.clientId);
+
+            assertEquals("logLevel should be equal before change", opts.logLevel, ablyRealtime.options.logLevel);
+            opts.logLevel = 33;
+            assertNotEquals("logLevel should not be equal after change", opts.logLevel, ablyRealtime.options.logLevel);
+
+            assertEquals("autoConnect should be equal before change", opts.autoConnect, ablyRealtime.options.autoConnect);
+            opts.autoConnect = true;
+            assertNotEquals("autoConnect should not be equal after change", opts.autoConnect, ablyRealtime.options.autoConnect);
+
+            assertEquals("logHandler should be equal before change", opts.logHandler, ablyRealtime.options.logHandler);
+            opts.logHandler = new Log.DefaultHandler();
+            assertNotEquals("logHandler should not be equal after change", opts.logHandler, ablyRealtime.options.logHandler);
```

Comment:
Instead of asserting 'not equal' cases, wouldn't it be better to assert equality with the original client options?

Reply:
I am not sure what you suggest. Please make a request change so we can all see it.

Reply:
What I suggest is that I see that you assert the manipulated client options is not going to be equal to the original one.
But what we care is that the original value hasn't changed at all - So let me split it in parts
* Get a copy of original options to compare against the ClientOptions later
* Manipulate ClientOptions' properties
* Compare that copy with ClientOptions in SDK and assert that they are equal

Reply:
I have added comparators with original values for every variable https://github.com/ably/ably-java/pull/827/commits/e9fa9409729ad4ea6b925f6fea118e9afbbfca24